### PR TITLE
Setup build targets and autogen for generating code from templates

### PR DESCRIPTION
**Pull request by xichen2020**
cc @robskillington @prateek @jeromefroe 

This PR adds the build targets for install the binary (`genny`) for code generation and sets up autogen to support code generation

### CODE REVIEW

File: auto-gen.sh [hunk 1]
```diff
@@ -99,6 +113,8 @@ set -e
 
 if [ "$2" = "generated/mocks" ]; then
     mocks_clear "*_mock.go"
```
Comment (jeromefroe) [inline auto-gen.sh]: Should this be `remove_matching_files ` now?

File: auto-gen.sh [hunk 2]
```diff
@@ -107,6 +123,8 @@ go generate $PACKAGE/$2
 
 if [ "$2" = "generated/mocks" ]; then
     mocks_cleanup "*_mock.go"
```
Comment (jeromefroe) [inline auto-gen.sh]: Same here?

File: auto-gen.sh (hunk 1 above)
Comment (xichen2020) [reply auto-gen.sh]: Yeah good catch ta, missed this during merge conflicts.

File: auto-gen.sh (hunk 2 above)
Comment (xichen2020) [reply auto-gen.sh]: This one actually is correct.

**Pull request merged**